File: native/provider-token-sharing-tests.swift
import Foundation

enum ProviderTokenSharingTests {
    private static func require(_ condition: Bool) throws {
        guard condition else { throw CocoaError(.fileReadCorruptFile) }
    }

    private static func status(reason: String, canEnable: Bool = false, token: Any = NSNull()) -> [String: Any] {
        ["version": 1, "enabled": false, "canEnable": canEnable, "reason": reason, "token": token]
    }

    static func parseSelfTest() throws {
        let source = try JSONSerialization.data(withJSONObject: status(reason: "source-unavailable"))
        try require(try QuotaSharingStatus.parse(source, channel: .providerTokens).reason == "source-unavailable")
        try require((try? QuotaSharingStatus.parse(source)) == nil)
        let account = try JSONSerialization.data(withJSONObject: status(reason: "account-unavailable"))
        try require(try QuotaSharingStatus.parse(account).reason == "account-unavailable")
        try require((try? QuotaSharingStatus.parse(account, channel: .providerTokens)) == nil)
        let token = String(repeating: "a", count: 64)
        let ready = status(reason: "ready", canEnable: true, token: token)
        let parsed = try QuotaSharingStatus.parse(JSONSerialization.data(withJSONObject: ready), channel: .providerTokens)
        try require(parsed.canEnable && parsed.token == token)

        var extra = ready
        extra["private"] = "must not pass"
        var missing = ready
        missing.removeValue(forKey: "token")
        var wrongVersion = ready
        wrongVersion["version"] = 2
        var wrongType = ready
        wrongType["enabled"] = "false"
        let invalid = [extra, missing, wrongVersion, wrongType,
            status(reason: "unknown"),
            status(reason: "ready", canEnable: true, token: NSNull()),
            status(reason: "source-unavailable", canEnable: true, token: token),
            status(reason: "source-unavailable", token: token),
            status(reason: "ready", canEnable: true, token: token + "\n"),
            status(reason: "ready", canEnable: true, token: String(repeating: "A", count: 64))]
        for object in invalid {
            let data = try JSONSerialization.data(withJSONObject: object)
            try require((try? QuotaSharingStatus.parse(data, channel: .providerTokens)) == nil)
        }
        try require((try? QuotaSharingStatus.parse(Data(repeating: 32, count: 4097), channel: .providerTokens)) == nil)
    }

    // The caller supplies packaged resources. All mutable data belongs to this
    // new, unpaired runtime, never the user's configured runtime.
    static func bridgeSelfTest(resources: URL) async throws {
        try parseSelfTest()
        let runtime = FileManager.default.temporaryDirectory.appendingPathComponent("observatory-provider-token-bridge-" + UUID().uuidString)
        try FileManager.default.createDirectory(at: runtime, withIntermediateDirectories: false)
        defer { try? FileManager.default.removeItem(at: runtime) }
        let current = try await QuotaSharing.run(runtime: runtime, resources: resources, action: "status", token: nil, channel: .providerTokens)
        try require(!current.enabled && !current.canEnable && current.reason == "pairing-unavailable" && current.token == nil)
        let disabled = try await QuotaSharing.run(runtime: runtime, resources: resources, action: "disable", token: nil, channel: .providerTokens)
        try require(!disabled.enabled && !disabled.canEnable && disabled.reason == "pairing-unavailable")
        var rejected = false
        do {
            _ = try await QuotaSharing.run(runtime: runtime, resources: resources, action: "enable", token: String(repeating: "a", count: 64), channel: .providerTokens)
        } catch { rejected = true }
        try require(rejected)
        try require(!FileManager.default.fileExists(atPath: runtime.appendingPathComponent("private-sync/provider-tokens.sqlite").path))
        try require(!FileManager.default.fileExists(atPath: runtime.appendingPathComponent("private-quota").path))
    }
}

File: native/windows/provider-token-sharing-tests.cs
using System.Text.Json.Nodes;

namespace WorkspaceObservatory;

internal static class ProviderTokenSharingTests
{
    private static void Require(bool condition)
    {
        if (!condition) throw new InvalidOperationException("Provider token sharing bridge test failed.");
    }

    private static JsonObject Status(string reason, bool canEnable = false, string? token = null) => new()
    {
        ["version"] = 1, ["enabled"] = false, ["canEnable"] = canEnable, ["reason"] = reason, ["token"] = token
    };

    private static bool Rejects(string text, SharingChannel channel)
    {
        try { QuotaSharing.Parse(text, channel); return false; }
        catch { return true; }
    }

    internal static void ParseSelfTest()
    {
        var source = Status("source-unavailable").ToJsonString();
        Require(QuotaSharing.Parse(source, SharingChannel.ProviderTokens).Reason == "source-unavailable");
        Require(Rejects(source, SharingChannel.Quota));
        var account = Status("account-unavailable").ToJsonString();
        Require(QuotaSharing.Parse(account).Reason == "account-unavailable");
        Require(Rejects(account, SharingChannel.ProviderTokens));
        var token = new string('a', 64);
        var ready = Status("ready", true, token);
        var parsed = QuotaSharing.Parse(ready.ToJsonString(), SharingChannel.ProviderTokens);
        Require(parsed.CanEnable && parsed.Token == token);

        var extra = (JsonObject)ready.DeepClone(); extra["private"] = "must not pass";
        var missing = (JsonObject)ready.DeepClone(); missing.Remove("token");
        var wrongVersion = (JsonObject)ready.DeepClone(); wrongVersion["version"] = 2;
        var wrongType = (JsonObject)ready.DeepClone(); wrongType["enabled"] = "false";
        foreach (var invalid in new[] { extra, missing, wrongVersion, wrongType,
            Status("unknown"), Status("ready", true), Status("source-unavailable", true, token),
            Status("source-unavailable", false, token), Status("ready", true, token + "\n"),
            Status("ready", true, new string('A', 64)) })
            Require(Rejects(invalid.ToJsonString(), SharingChannel.ProviderTokens));
        Require(Rejects(new string(' ', 4097), SharingChannel.ProviderTokens));
        Require(Rejects(source, (SharingChannel)999));
    }

    // The executable's packaged resources are reused, but all mutable data is
    // confined to a new unpaired runtime instead of the user's runtime.
    internal static async Task BridgeSelfTest()
    {
        ParseSelfTest();
        var runtime = Path.Combine(Path.GetTempPath(), "observatory-provider-token-bridge-" + Guid.NewGuid().ToString("N"));
        Directory.CreateDirectory(runtime);
        try
        {
            var current = await QuotaSharing.Run(runtime, "status", null, CancellationToken.None, SharingChannel.ProviderTokens);
            Require(!current.Enabled && !current.CanEnable && current.Reason == "pairing-unavailable" && current.Token is null);
            var disabled = await QuotaSharing.Run(runtime, "disable", null, CancellationToken.None, SharingChannel.ProviderTokens);
            Require(!disabled.Enabled && !disabled.CanEnable && disabled.Reason == "pairing-unavailable");
            var rejected = false;
            try { await QuotaSharing.Run(runtime, "enable", new string('a', 64), CancellationToken.None, SharingChannel.ProviderTokens); }
            catch (InvalidOperationException) { rejected = true; }
            Require(rejected);
            Require(!File.Exists(Path.Combine(runtime, "private-sync", "provider-tokens.sqlite")));
            Require(!Directory.Exists(Path.Combine(runtime, "private-quota")));
        }
        finally { Directory.Delete(runtime, recursive: true); }
    }
}
